File: backend/domain/UnlockRules.cpp
#include "domain/UnlockRules.h"

#include <algorithm>

namespace wm {

std::map<NodeId, NodeState> UnlockRules::derive(const SkillTree& tree, const Progress& progress) {
  std::map<NodeId, NodeState> states;
  for (const NodeSpec& node : tree.nodes()) {
    if (progress.completed.count(node.id)) {
      states[node.id] = NodeState::complete;
      continue;
    }
    if (progress.inProgress.count(node.id)) {
      states[node.id] = NodeState::active;
      continue;
    }
    bool unlocked = std::all_of(node.prerequisites.begin(), node.prerequisites.end(),
                                [&](const NodeId& prereq) { return progress.completed.count(prereq) > 0; });
    states[node.id] = unlocked ? NodeState::available : NodeState::locked;
  }
  return states;
}

}

File: web/src/products/roadmap/model/UnlockRules.js
// The only place that derives NodeState; nothing else may hand-set a node state.

export class UnlockRules {
  static derive(tree, progress) {
    const states = new Map();

    for (const node of tree.nodes) {
      if (progress.completed.has(node.id)) {
        states.set(node.id, 'complete');
        continue;
      }
      if (progress.inProgress.has(node.id)) {
        states.set(node.id, 'active');
        continue;
      }
      const unlocked = node.prerequisites.every((prereqId) => progress.completed.has(prereqId));
      states.set(node.id, unlocked ? 'available' : 'locked');
    }

    return states;
  }
}
